### PR TITLE
Failure Detector

### PR DESCRIPTION
**Pull request by zalokhan**
Additions:
- Management Server
  - 4th Server launched from Corfu Server with the Layout, Logunit and Sequencer.
  - Schedules a thread to detect failures using the desired policy.
  - Policy used here is periodic polling
  - Once a failure is detected, the failureHandlerDispatcher is triggered.

- FailureHandler
  - First seals the epoch to prevent access to the old layout.
  - Generates the new layout by removing failed layout servers.
  - Runs the paxos algorithm on the new layout and commits it.

---
This change is [Reviewable](https://reviewable.io/reviews/corfudb/corfudb/242)

### CODE REVIEW

**Comment by no2chem** [comment]
Addresses #237

**Comment by no2chem** [comment]
Review status: 0 of 12 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/main/java/org/corfudb/infrastructure/CorfuServer.java, line 167 at r1](https://reviewable.io:443/reviews/corfudb/corfudb/242#-KTw_0oxc1U1c2v-de62:-KTw_0oykce4inuWt2mz:bolrvgd) ([raw file](https://github.com/corfudb/corfudb/blob/e608428aa6a26aeb5a0793ab6124a51eaeb905af/src/main/java/org/corfudb/infrastructure/CorfuServer.java#L167)):*
> ```Java
>         router.addServer(layoutServer);
>         router.addServer(new LogUnitServer(serverContext));
>         router.addServer(new ManagementServer(serverContext, new PeriodicPollPolicy(), layoutServer));
> ```

Could you make PollPolicy a member of serverContext and pass the policy via tghe context?

---

*[src/main/java/org/corfudb/runtime/clients/IClientRouter.java, line 125 at r1](https://reviewable.io:443/reviews/corfudb/corfudb/242#-KTw_TX9ERreVxBlAt7h:-KTw_TX9ERreVxBlAt7i:bxrtfmh) ([raw file](https://github.com/corfudb/corfudb/blob/e608428aa6a26aeb5a0793ab6124a51eaeb905af/src/main/java/org/corfudb/runtime/clients/IClientRouter.java#L125)):*
> ```Java
> 
>     default void setTimeoutConnect(long timeoutConnect) {
> ```

since these default members actually are implemented in the test/netty routers, there is no reason to make them default.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/corfudb/corfudb/242)*

**Comment by zalokhan** [comment]
Review status: 0 of 17 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/main/java/org/corfudb/infrastructure/CorfuServer.java, line 167 at r1](https://reviewable.io:443/reviews/corfudb/corfudb/242#-KTw_0oxc1U1c2v-de62:-KU3HW7LoeZPxKfr9HtI:b-896fix) ([raw file](https://github.com/corfudb/corfudb/blob/e608428aa6a26aeb5a0793ab6124a51eaeb905af/src/main/java/org/corfudb/infrastructure/CorfuServer.java#L167)):*
<details><summary><i>Previously, no2chem (Michael Wei) wrote…</i></summary>
> Could you make PollPolicy a member of serverContext and pass the policy via tghe context?
</details>
Done.

---

*[src/main/java/org/corfudb/runtime/clients/IClientRouter.java, line 125 at r1](https://reviewable.io:443/reviews/corfudb/corfudb/242#-KTw_TX9ERreVxBlAt7h:-KU40QByl_GwKu5Ajr9Q:b-896fix) ([raw file](https://github.com/corfudb/corfudb/blob/e608428aa6a26aeb5a0793ab6124a51eaeb905af/src/main/java/org/corfudb/runtime/clients/IClientRouter.java#L125)):*
<details><summary><i>Previously, no2chem (Michael Wei) wrote…</i></summary>
> since these default members actually are implemented in the test/netty routers, there is no reason to make them default.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/corfudb/corfudb/242)*

**Comment by no2chem** [comment]
Review status: 0 of 17 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[src/main/java/org/corfudb/runtime/CorfuRuntime.java, line 237 at r2](https://reviewable.io:443/reviews/corfudb/corfudb/242#-KUGWvhf30OObZhL_6Mc:-KUGWvhf30OObZhL_6Md:bnmapgk) ([raw file](https://github.com/corfudb/corfudb/blob/097ecf9e4dc9a847dbb24913770214e56c15c2b4/src/main/java/org/corfudb/runtime/CorfuRuntime.java#L237)):*
> ```Java
>         CompletableFuture<Layout> l = fetchLayout();
>         if (l == null) {
>             System.out.println("O NO");
> ```

might want to take these old debug statements out :)

---

*[src/main/java/org/corfudb/runtime/CorfuRuntime.java, line 269 at r2](https://reviewable.io:443/reviews/corfudb/corfudb/242#-KUGWzzCHYpYrdISs-iA:-KUGWzzCHYpYrdISs-iB:b60ey12) ([raw file](https://github.com/corfudb/corfudb/blob/097ecf9e4dc9a847dbb24913770214e56c15c2b4/src/main/java/org/corfudb/runtime/CorfuRuntime.java#L269)):*
> ```Java
>                     // with  a null runtime.
>                     if (layoutFuture == null) {
>                         System.out.println("O NO2");
> ```

and this one too

---

*[src/test/java/org/corfudb/runtime/view/LayoutViewTest.java, line 150 at r2](https://reviewable.io:443/reviews/corfudb/corfudb/242#-KUGX2HCbz5mJ0sF8NpU:-KUGX2HCbz5mJ0sF8NpV:b-eytxsb) ([raw file](https://github.com/corfudb/corfudb/blob/097ecf9e4dc9a847dbb24913770214e56c15c2b4/src/test/java/org/corfudb/runtime/view/LayoutViewTest.java#L150)):*
> ```Java
>      */
> //    @Test
> //    public void reconfigurationDuringDataOperations()
> ```

Is there a reason this test case is commented out?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/corfudb/corfudb/242)*

**Comment by zalokhan** [comment]
Oh yes Im sorry about those. Thought I had removed those.
Will fix that.

**Comment by no2chem** [comment]
Hi Zeeshan, 

Is there any status on getting this branch merged against master?

**Comment by zalokhan** [comment]
Yes Michael there are a few issues in the tests.
Im trying to figure out.
For example:
https://github.com/CorfuDB/CorfuDB/pull/242/files#diff-a4dbee403d59698840625da1df0fd42eR49
This variable here shouldn't be static as the tests share this file and mess up the management server operations.
Ill then rebase it with the latest master commit and update you.

File: src/main/java/org/corfudb/infrastructure/ManagementServer.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    private static boolean isBootstrapped = false;
+
+    /**
+     * A scheduler, which is used to schedule failure detection.
+     */
+    private final ExecutorService schedulerService =
+            Executors.newSingleThreadExecutor(
+                    new ThreadFactoryBuilder()
+                            .setDaemon(true)
+                            .setNameFormat("Config-Mgr-%d")
```
Comment (Maithem) [inline src/main/java/org/corfudb/infrastructure/ManagementServer.java]: Please use a more appropriate name.

Also, we shouldn't allocate a thread per monitored host. In other words, we want a couple of threads monitoring 1000s of hosts as opposed to 1000s of threads.

File: src/main/java/org/corfudb/infrastructure/ManagementServer.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    /**
+     * Wait until we receive a message with the layout endpoint to bootstrap the management server.
+     */
+    private void waitForBootstrap() {
+        Layout currentLayout = layoutServer.getCurrentLayout();
+        while (currentLayout == null) {
+            try {
+                Thread.sleep(1000);
```
Comment (Maithem) [inline src/main/java/org/corfudb/infrastructure/ManagementServer.java]: These sleeps seem arbitrary, what happens if the sleep needs to be change?

Comment (zalokhan) [inline src/main/java/org/corfudb/infrastructure/ManagementServer.java]: I'll replace this with a configurable retry variable.

**Comment by zalokhan** [comment]
https://github.com/CorfuDB/CorfuDB/pull/242/files#diff-a4dbee403d59698840625da1df0fd42eR83
This while loop was causing the issue.
A lot of test cases would finish early and since I am not handling the Interrupt properly these threads would never die and are stuck in this loop forever.

**Comment by zalokhan** [comment]
We can close this PR as I am breaking it up into multiple smaller PRs.

**Comment by slfritchie** [comment]
@zalokhan If you had meant to close this PR, please go ahead and do it now.

**Pull request closed without merging**